File: basicAlgo/mergeSort.py


array = [10,9,8,7,6,5,4,3,2,1,11,12,13,16,15,14,17,19,18,20]

n = len(array)

def merge(leftArray,rightArray):
    
    newArray = []
    leftArrayIndex = 0
    rightArrayIndex = 0
    
    while len(leftArray) > leftArrayIndex and len(rightArray) > rightArrayIndex:
        if leftArray[leftArrayIndex] < rightArray[rightArrayIndex]:
            newArray.append(leftArray[leftArrayIndex])
            leftArrayIndex += 1
        else:
            newArray.append(rightArray[rightArrayIndex])
            rightArrayIndex += 1
    
    while len(leftArray) > leftArrayIndex:
        newArray.append(leftArray[leftArrayIndex])
        leftArrayIndex += 1
        
    while len(rightArray) > rightArrayIndex:
        newArray.append(rightArray[rightArrayIndex])
        rightArrayIndex += 1
        
    return newArray

def mergeSort(array,left,right):
    
    if left == right:
        return [array[left]]
    
    mid = (left + right) // 2
    
    leftArray = mergeSort(array,left,mid) 
    rightArray = mergeSort(array,mid+1,right)
    
    return merge(leftArray,rightArray)
        
sortedArray = mergeSort(array,0,n-1)

print(sortedArray)

File: basicAlgo/mergeSort.js
const merge = (array1, array2) => {
  const mergeSortArray = [];

  let i = 0;
  let j = 0;

  while (i < array1.length && j < array2.length) {
    if (array1[i] > array2[j]) {
      mergeSortArray.push(array2[j]);
      j++;
    } else {
      mergeSortArray.push(array1[i]);
      i++;
    }
  }

  while (j < array2.length) {
    mergeSortArray.push(array2[j]);
    j++;
  }

  while (i < array1.length) {
    mergeSortArray.push(array1[i]);
    i++;
  }

  return mergeSortArray;
};

const mergeSort = (array, start, end) => {
  if (start === end) return [array[end]];

  const mid = Math.floor((end + start) / 2);

  const mergeSortArray1 = mergeSort(array, start, mid);
  const mergeSortArray2 = mergeSort(array, mid + 1, end);

  return merge(mergeSortArray1, mergeSortArray2);
};

const array = [5, 4, 3, 2, 1];

const sortedArray = mergeSort(array, 0, array.length - 1);

console.log(array);
console.log(sortedArray);
